packages/node/src/indexer/unfinalizedBlocks.service.ts

```diff
@@ -0,0 +1,219 @@
+  }
+
+  // remove any records less and equal than input finalized blockHeight
+  removeFinalized(blockHeight: number): void {
+    Object.entries(this.unfinalizedBlocks).map(([bestBlockHeight, hash]) => {
+      if (Number(bestBlockHeight) <= blockHeight) {
+        delete this.unfinalizedBlocks[bestBlockHeight];
+      }
+    });
+  }
+
+  // find closest record from block heights
+  getClosestRecord(
+    blockHeight: number,
+  ): { blockHeight: number; hash: HexString } | undefined {
+    // Have the block in the best block, can be verified
+    if (Object.keys(this.unfinalizedBlocks).length !== 0) {
+      const record = this.getSortedUnfinalizedBlocks().find(
+        ([bestBlockHeight, hash]) => Number(bestBlockHeight) <= blockHeight,
+      );
+      if (record) {
+        const [bestBlockHeight, hash] = record;
+        return { blockHeight: Number(bestBlockHeight), hash: hash };
+      }
+    }
+    return undefined;
+  }
+
+  // verify best blocks with finalized block
+  async validateUnfinalizedBlocks(): Promise<boolean> {
+    const finalizedBlockNumber =
+      this.finalizedBlock.block.header.number.toNumber();
+    const lastVerifiableBlock = this.getClosestRecord(finalizedBlockNumber);
+    if (lastVerifiableBlock) {
+      if (lastVerifiableBlock.blockHeight === finalizedBlockNumber) {
+        if (lastVerifiableBlock.hash !== this.finalizedBlock.hash.toHex()) {
+          return false;
+        }
+      } else {
+        // use api to fetch lastVerifiableBlock, see if hash matches
+        const actualHash = (
+          await this.api.rpc.chain.getBlockHash(lastVerifiableBlock.blockHeight)
+        ).toHex();
+        if (actualHash !== lastVerifiableBlock.hash) {
+          logger.warn(
+            `Block fork found, enqueued un-finalized block at ${lastVerifiableBlock.blockHeight} with hash ${lastVerifiableBlock.hash}, actual hash is ${actualHash} `,
+          );
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  async getLastCorrectFinalizedBlock(): Promise<number | undefined> {
+    const bestVerifiableBlocks = this.getSortedUnfinalizedBlocks()
+      .filter(
+        ([bestBlockHeight, hash]) =>
+          Number(bestBlockHeight) <=
+          this.finalizedBlock.block.header.number.toNumber(),
+      )
+      .reverse(); // We work backwards from finalized head
+
+    let checkingHeader = this.finalizedBlock.block.header;
+
+    // Work backwards through the blocks until we find a matching hash
+    for (const [block, hash] of bestVerifiableBlocks) {
+      // Should match for index 0
+      if (hash === this.finalizedBlock.block.header.hash.toHex()) {
+        return Number(block);
+      }
+      // Should match for index 1....n
+      if (hash === checkingHeader.parentHash.toHex()) {
+        return Number(block);
+      }
+
+      // Get the new parent
+      checkingHeader = await this.api.rpc.chain.getHeader(
```

Comment:
How many times you need to search backwards isn't decided by length of bestVerifiableBlocks, but the distance of finalised block height and unfinalised block height.